src/python/pants/backend/cc/dependency_inference/rules.py

```diff
@@ -65,10 +65,44 @@ class CCFilesMapping:
     mapping: FrozenDict[str, Address]
     ambiguous_files: FrozenDict[str, tuple[Address, ...]]
     mapping_not_stripped: FrozenDict[str, Address]
+    mapping_relative_to_include_dirs: FrozenDict[str, Address]
+
+
+def _map_relative_to_include_directories(
+    cc_file_mapping: dict[str, Address], include_directories: Iterable[str]
+) -> dict[str, Address]:
+    """Returns a mapping of CC files (made relative to a potential set of include_directories) to
+    their owning file address.
+
+    This is used to infer dependencies when an #include statement is relative to an include_directory, but does
+    not explicitly state that relative include_directory.
+
+    For example, a project with a file at `root/include/mylib/foo.h` may be referenced via `#include "mylib/foo.h"`,
+    with pre-defined build settings to reference the `root/include` automatically.
+
+    This is a common pattern in C/C++ libraries, where the public headers are all in a single directory (e.g. `include/`)
+    and the library is compiled with `-Iinclude` to automatically reference the public headers. Private headers are
+    typically in a separate directory alongside source files (e.g. `src/`).
+
+    TODO: Would a more "correct" implementation be to grab SourceRoots and use those to determine the relative path? Currently,
+    this implementation progresses up the file's parent chain until it finds a directory that matches an include_directory, and then
+    uses that as the relative path.
```

Comment:
Yeah, so that's what I did in the original PR: https://github.com/pantsbuild/pants/pull/16424/files#diff-6c96e874c9a3590582c7988d29f8584d67fb5da102218a4ad5da320d8143f5bc

So, the preferred version is searching relative to `{source_root}/{include}` instead of going up the parent chain until we hit `{include}`? Perfectly happy either way, since I can contrive examples that would break both, so 6 of one 🤷🏽 

The underlying assumption here is that (at least) the top-level folder of the library/module is a `source_root`. If not, the whole thing falls over. If that's a valid precondition to enforce, I'm good. C/C++ doesn't really have a standard marker that would define a lib root, unfortunately - so it would need to be explicitly placed by the user.

Reply:
Source roots can also be used for namespacing decisions (at least in Python). How would C/C++ code be impacted by having a source root for the include directory but still have source code "above" an include directory in a parent directory referencing those headers? Such parent directories are outside the source root for the include directory and would need to be in there own source root, right?

Reply:
So you'd need to make `foo` and `foo/include` be source roots versus just making `foo` be the source root because the `foo/include` directory can be viewed as associated with the `foo` source root.

Reply:
In my very limited world view, I would want/expect something like:

```
root/mylib/SOURCE_ROOT
root/mylib/BUILD -> cc_library(name=mylib, headers=, ...)
root/mylib/include/mylib/foo.h
root/mylib/src/foo.c
```

```
root/myexe/SOURCE_ROOT
root/myexe/BUILD -> cc_binary(name=, deps=[mylib] ...)
root/myexe/src/bar.h
root/myexe/src/main.c -> "#include mylib/foo.h"
```

And if `mylib` wasn't specified in the `deps`, but we allowed discovering `mylib/foo.h` across source roots, then the `mylib` library would automatically be included, so that we correctly access headers + objects (and back-compile as necessary)

Reply:
I think in the cc context, "source roots" only means "include paths"? so `foo/include` is a source root but `foo` is not? Trying to think why we would need it to be.

Reply:
So I kinda do think this should use source roots, since that is a concept we have already invested in, and documented.

Reply:
I guess it's kinda interesting, because I use Source Roots synonymously with project roots (because that's how all my stuff ends up being namespaced). 

However, if there is buy-in for source-roots being synonymous with include_directories, then I think that would be fine - but is there an equivalent concept of project root? Like, where a `package.json` or the first `CMakeLists.txt` would be located?

So, I guess we'd need something like this?

```python
[source]
root_patterns = [
    "/root/mylib1/include",
    "/root/mylib2/include",
]
```

Reply:
🤷🏽 Probably doesn't - I use it for building stuff relative to a project root (in a monorepo), and ensuring dep inference doesn't escape that project root - to make sure everything is handled through my public cc_library targets (for example).

So, I'll happily call that an SJ thing, not a Pants thing - and I'll probably enforce it using visibility going forward.

Reply:
> to know about a distinct project-root concept

Aside: Doesn't Go kinda use this? Swift would, definitely - it's the location of the Package.swift, and everything is relative to it.

For this ticket, I think I'll just scrap all my changes - but keep my extra tests, since Tom's original dep inference already covered directories with correct source roots (as mentioned in the PR description "An alternative option is to specify mylib/include as a source root")

Reply:
The Go backend doesn't use source roots at all. It infers Go import paths based on where a `go_package` target is relative to its "owning" `go_mod` target (just as the `go` tool does).

Reply:
So yes, the Go backend has a notion of "project root" based on the location of the `go_mod` targets.

Reply:
Ahhhh, okay okay - yeah, so that's what Swift would do to. Thanks for clearing that up.

This is sounding more and more like an unnecessary PR then, with the CC docs needing to point out that source roots will essentially operate the same as include_directories would, and then call it a day.

Reply:
@tdyas You were right about the source files source root - I didn't realize StrippedFileNames requires a source root.

Reply:
So, that adds another wrinkle, I think - because relative paths shouldn't need the source roots (if they're acting as include directories), I don't think... 

Going to need to re-think this a bit - so that gcc/clang semantics also line up with what Pants expects a bit better.

Reply:
Not sure I understand "relative paths shouldn't need the source roots"? if in `foo/bar.cc` you `#include baz/qux.h` the preprocessor will first look for `foo/baz/qux.h`, and then in `dir/baz/qux.h` for every dir provided via `-I`, and then for every dir in the system include dirs. If `foo/baz/qux.h` exists then it is easy to infer that it is the right file, no? If not, that's when you look on the source roots.

Reply:
It was just a comment on the fact that the `StrippedFileNames` API uses source roots, which I didn't know. That API was used already in the dep inference. Tangential to the include_directories discussion.